File: 2023/13/solution.js
const fs = require("fs");
const path = require("path");

const input = fs
  .readFileSync(path.join(__dirname, "input.txt"), "utf-8")
  .split("\n\n")
  .map((a) => a.split("\n"));

class Pattern {
  constructor(input) {
    this.pattern = input;
    [this.reflectionType, this.reflectionRow, this.reflectionCol] =
      this.findReflection();
    [this.smudgeType, this.smudgeRow, this.smudgeCol] = this.findSmudge();
  }

  findReflection() {
    for (let col = 1; col < this.pattern[0].length; col++) {
      let reflection = true;
      for (let row = 0; row < this.pattern.length; row++) {
        for (let i = 0; i < col && i + col < this.pattern[0].length; i++) {
          if (this.pattern[row][col - i - 1] !== this.pattern[row][i + col]) {
            reflection = false;
            break;
          }
        }
        if (!reflection) break;
      }
      if (reflection) return ["col", 0, col];
    }
    for (let row = 1; row < this.pattern.length; row++) {
      let reflection = true;
      for (let col = 0; col < this.pattern[0].length; col++) {
        for (let i = 0; i < row && i + row < this.pattern.length; i++) {
          if (this.pattern[row - i - 1][col] !== this.pattern[i + row][col]) {
            reflection = false;
            break;
          }
        }
        if (!reflection) break;
      }
      if (reflection) return ["row", row, 0];
    }
    return [null, 0, 0];
  }

  findSmudge() {
    for (let col = 1; col < this.pattern[0].length; col++) {
      let smudges = 0;
      for (let row = 0; row < this.pattern.length; row++) {
        for (let i = 0; i < col && i + col < this.pattern[0].length; i++) {
          if (this.pattern[row][col - i - 1] !== this.pattern[row][i + col]) {
            smudges++;
            if (smudges > 1) break;
          }
        }
        if (smudges > 1) break;
      }
      if (smudges === 1) return ["col", 0, col];
    }
    for (let row = 1; row < this.pattern.length; row++) {
      let smudges = 0;
      for (let col = 0; col < this.pattern[0].length; col++) {
        for (let i = 0; i < row && i + row < this.pattern.length; i++) {
          if (this.pattern[row - i - 1][col] !== this.pattern[i + row][col]) {
            smudges++;
            if (smudges > 1) break;
          }
        }
        if (smudges > 1) break;
      }
      if (smudges === 1) return ["row", row, 0];
    }
    return [null, 0, 0];
  }
}

const patterns = input.map((a) => new Pattern(a));
const part1 = patterns.reduce(
  (a, b) => a + 100 * b.reflectionRow + b.reflectionCol,
  0
);
console.log(part1);

const part2 = patterns.reduce((a, b) => a + 100 * b.smudgeRow + b.smudgeCol, 0);
console.log(part2);

File: 2023/13/solution.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strings"
)

func main() {
	_, filename, _, _ := runtime.Caller(0)
	dirname := filepath.Dir(filename)
	inputFilePath := filepath.Join(dirname, "input.txt")
	data, err := os.ReadFile(inputFilePath)
	if err != nil {
		panic(err)
	}
	puzzleInput := parseInput(strings.Split(string(data), "\n\n"))
	fmt.Println(part1(puzzleInput))
	fmt.Println(part2(puzzleInput))
}

type Pattern []string

func (p Pattern) FindReflection(tolerance int) (int, int) {
	var smudges int
	for r := 1; r < len(p); r++ {
		smudges = 0
		for dr := 1; r-dr >= 0 && r+dr-1 < len(p); dr++ {
			for c := range p[0] {
				if p[r-dr][c] != p[r+dr-1][c] {
					smudges++
					if smudges > tolerance {
						break
					}
				}
			}
			if smudges > tolerance {
				break
			}
		}
		if smudges == tolerance {
			return r, 0
		}
	}
	for c := 1; c < len(p[0]); c++ {
		smudges = 0
		for dc := 1; c-dc >= 0 && c+dc-1 < len(p[0]); dc++ {
			for r := range p {
				if p[r][c-dc] != p[r][c+dc-1] {
					smudges++
					if smudges > tolerance {
						break
					}
				}
			}
			if smudges > tolerance {
				break
			}
		}
		if smudges == tolerance {
			return 0, c
		}
	}

	return -1, -1
}

func parseInput(data []string) []Pattern {
	patterns := make([]Pattern, len(data))
	for i, pattern := range data {
		patterns[i] = Pattern(strings.Split(pattern, "\n"))
	}
	return patterns
}

func part1(patterns []Pattern) int {
	res := 0
	for _, pattern := range patterns {
		r, c := pattern.FindReflection(0)
		if r == -1 {
			panic(fmt.Sprintf("Reflection not found for pattern:\n%s", strings.Join(pattern, "\n")))
		}
		res += 100*r + c
	}
	return res
}

func part2(patterns []Pattern) int {
	res := 0
	for _, pattern := range patterns {
		r, c := pattern.FindReflection(1)
		if r == -1 {
			panic(fmt.Sprintf("Smudged reflection not found for pattern:\n%s", strings.Join(pattern, "\n")))
		}
		res += 100*r + c
	}
	return res
}
